graspologic/match/solver.py

```diff
@@ -0,0 +1,734 @@
+# Copyright (c) Microsoft Corporation and contributors.
+# Licensed under the MIT License.
+
+import time
+import warnings
+from functools import wraps
+from typing import Callable, Optional
+
+import numpy as np
+from beartype import beartype
+from numba import njit
+from ot import sinkhorn
+from packaging import version
+from scipy import __version__ as scipy_version
+from scipy.optimize import linear_sum_assignment
+from scipy.sparse import csr_matrix
+from sklearn.utils import check_scalar
+
+if version.parse(scipy_version) >= version.parse("1.8.0"):
+    from scipy.sparse import csr_array
+else:
+    # HACK: what should this be?
+    csr_array = csr_matrix
+
+
+from graspologic.types import List, RngType, Tuple
+
+from .types import AdjacencyMatrix, Int, MultilayerAdjacency, PaddingType, Scalar
+
+
+def parametrized(dec: Callable) -> Callable:
+    def layer(*args, **kwargs) -> Callable:  # type: ignore
+        def repl(f: Callable) -> Callable:
+            return dec(f, *args, **kwargs)
+
+        return repl
+
+    return layer
+
+
+@parametrized
+def write_status(f: Callable, msg: str, level: int) -> Callable:
+    @wraps(f)
+    def wrap(*args, **kw):  # type: ignore
+        obj = args[0]
+        verbose = obj.verbose
+        if level <= verbose:
+            total_msg = (level - 1) * "   "
+            total_msg += obj.status() + " " + msg
+            print(total_msg)
+        if (verbose >= 4) and (level <= verbose):
+            ts = time.time()
+            result = f(*args, **kw)
+            te = time.time()
+            sec = te - ts
+            output = total_msg + f" took {sec:.3f} seconds."
+            print(output)
+        else:
+            result = f(*args, **kw)
+        return result
+
+    return wrap
+
+
+class GraphMatchSolver:
```

Comment:
quick note: unless you're meaning to make all of the member variables of GraphMatchSolver public and mutable by the world, I'd at _least_ suggest having the member variables be prepended with a `_`, ala `self._init`.  If you do mean for them to be public, that's fine; just saying it's python convention to prepend with `_` to denote that it's not part of the public api

Reply:
i think the entire solver should probably be _GraphMatchSolver i.e. private.

if I make that change - should all fields/methods still have that leading underscore?

Reply:
I think it's fair to say no; you already said the class is not part of the public api, so ... I'm asserting the rest should follow suit?

Reply:
err, to clarify: don't _ the member variables if the class is _.  unless you feel like testing out the symbol rename feature in your IDE of choice